src/Billing/BillingProcessor/X12RemoteTracker.php

```diff
@@ -0,0 +1,222 @@
+    const STATUS_LOGIN_ERROR = 'login-error';
+    const STATUS_CHDIR_ERROR = 'chdir-error';
+    const STATUS_IN_PROGRESS = 'in-progress';
+    const STATUS_UPLOAD_ERRROR = 'upload-error';
+    const STATUS_SUCCESS = 'success';
+
+    const TABLE_NAME = 'x12_remote_tracker';
+
+    protected static $x12_partner_field_keys = [
+        'x12_sftp_host' => 'X12 SFTP Host',
+        'x12_sftp_port' => 'X12 SFTP Port',
+        'x12_sftp_login' => 'X12 SFTP Login',
+        'x12_sftp_pass' => 'X12 SFTP Password',
+        'x12_sftp_remote_dir' => 'X12 SFTP Remote Dir',
+        'x12_sftp_local_dir' => 'X12 SFTP Local Dir',
+    ];
+
+    const SELECT = "SELECT R.id, R.x12_filename, R.status, R.messages, R.claims, R.created_at, R.updated_at,
+       P.name, P.id AS x12_partner_id, P.x12_sftp_host, P.x12_sftp_port, P.x12_sftp_login, P.x12_sftp_pass,
+       P.x12_sftp_remote_dir, P.x12_sftp_local_dir FROM x12_remote_tracker R";
+
+    protected $validationMessages = [];
+
+    public function __construct()
+    {
+        parent::__construct(self::TABLE_NAME);
+    }
+
+    public static function sftpSendWaitingFiles()
+    {
+        $remoteTracker = new X12RemoteTracker();
+        $x12_remotes = $remoteTracker->fetchByStatus(self::STATUS_WAITING);
+        $x12_remote['messages'] = [];
+        foreach ($x12_remotes as $x12_remote) {
+
+            // Make sure required parameters are filled in on the X12 partner form, otherwise, log a message
+            if (false === $remoteTracker->validateSFTPCredentials($x12_remote)) {
+                // there was a problem, get messages, log them and continue
+                $x12_remote['status'] = self::STATUS_PARAMETER_ERROR;
+                $x12_remote['messages'] = array_merge($x12_remote['messages'], $remoteTracker->validationMessages);
+                $remoteTracker->update($x12_remote);
+                continue;
+            }
+
+            // Make sure local claim file exists and can we have permission to read it
+            // We try both the SFTP directory and the edi root directry
+            $claim_file = $x12_remote['x12_sftp_local_dir'] . $x12_remote['x12_filename'];
+            if (!file_exists($claim_file)) {
+                $claim_file = $GLOBALS['OE_SITE_DIR'] . "/documents/edi/" . $x12_remote['x12_filename'];
+            }
+
+            $claim_file_contents = file_get_contents($claim_file);
+            if (false === $claim_file_contents) {
+                $x12_remote['status'] = self::STATUS_CLAIM_FILE_ERROR;
+                $x12_remote['messages'][]= "Could not open local claim file: `$claim_file`";
+                $remoteTracker->update($x12_remote);
+                continue;
+            }
+
+            // Attempt to login
+            $sftp = new SFTP($x12_remote['x12_sftp_host'], $x12_remote['x12_sftp_port']);
+            if (false === $sftp->login($x12_remote['x12_sftp_login'], $x12_remote['x12_sftp_pass'])) {
```

Comment:
note that the x12 ftp password should be stored in the database in encrypted form